### ISSUE TITLE
gmock/gtest not found

### ISSUE THREAD

**Issue by cugone** [lairworks/nas2d-core#184]
Pulled latest on master and when building gmock and gtest headers can't be found.

**Comment by DanRStevens** [comment]
I switched to install Google Test (and Google Mock) with `vcpkg` instead of NuGet. You'd have to run the `vcpkg` install commands locally yourself. It's not as integrated as NuGet which will auto restore/update packages.

The `vcpkg` install commands from .appveyor.yml are:
```
vcpkg install gtest:x86-windows
vcpkg install gtest:x64-windows
```

**Comment by cugone** [comment]
That fixed it.

**Comment by cugone** [comment]
@DanRStevens 
@ldicker83 

Recently I've noticed that when I try to build the Test project it fails with the error `gtest.lib not found`.

I've found a fix for this but I'm not sure if it will affect AppVeyor.

In the properties page of the test project: `Linker > Input > Additional Dependencies` instead of listing `gtest.lib;gtest_main.lib` as local to the project adding the relative paths as `$(vcpkgroot)/lib/gtest.lib;$(vcpkgroot)/lib/manual-link/gtest_main.lib` allows the project to build.

However when run in Debug no tests are run, giving the following output (See below).

It's slightly better in Release. Although, the main project doesn't build the tests at least run but half the `FilesystemTest` tests fail. For the main project there is a conflict with the Code Analysis Rule set in use. It is a hard-coded path that does not exist locally. Disabling Code Analysis fixes the issue but I don't know if you want to do this.

For the tests that fail it seems to be because the initialization step of the `FilesystemTest` object sets the working directory to `%APPDATA%\Lairworks\NAS2DUnitTests` (it creates those folders) but the tests themselves expect the working directory to be local to the project so they never find the `data/` folder they are looking for.

Any thoughts?

This is the output when trying to run the tests in Debug:

> [5/23/2019 4:49:24 PM Informational] ------ Discover test started ------
[5/23/2019 4:49:25 PM Informational] ========== Discover test finished: 0 found (0:00:01.3580879) ==========
[5/23/2019 4:49:40 PM Informational] ------ Discover test started ------
[5/23/2019 4:49:40 PM Informational] ========== Discover test finished: 0 found (0:00:00.4495447) ==========
[5/23/2019 4:56:28 PM Informational] ------ Discover test started ------
[5/23/2019 4:56:28 PM Informational] ========== Discover test finished: 0 found (0:00:00.5338749) ==========
[5/23/2019 4:56:43 PM Informational] ------ Discover test started ------
[5/23/2019 4:56:43 PM Informational] ========== Discover test finished: 0 found (0:00:00.4539899) ==========
[5/23/2019 4:58:55 PM Informational] ------ Discover test started ------
[5/23/2019 4:58:56 PM Warning] Could not locate debug symbols for 'E:\Git\nas2dfork\test\x86\Debug\test.exe'. To make use of '--list_content' discovery, ensure that debug symbols are available or make use of '<ForceListContent>' via a .runsettings file.
[5/23/2019 4:58:56 PM Informational] Test Adapter for Google Test: Test discovery starting...
[5/23/2019 4:58:56 PM Error] ERROR: Could not list test cases for executable 'E:\Git\nas2dfork\test\x86\Debug\test.exe': process execution failed with exit code -1073741515
Command executed: 'E:\Git\nas2dfork\test\x86\Debug\test.exe --gtest_list_tests', working directory: 'E:\Git\nas2dfork\test\x86\Debug'
Command produced no output
[5/23/2019 4:58:56 PM Informational] Found 0 tests in executable E:\Git\nas2dfork\test\x86\Debug\test.exe
[5/23/2019 4:58:56 PM Informational] Test discovery completed, overall duration: 00:00:00.1339936
[5/23/2019 4:58:56 PM Error] ERROR: 
================
The following errors and warnings have occured during test discovery  (enable debug mode for more information):
ERROR: Could not list test cases for executable 'E:\Git\nas2dfork\test\x86\Debug\test.exe': process execution failed with exit code -1073741515
Command executed: 'E:\Git\nas2dfork\test\x86\Debug\test.exe --gtest_list_tests', working directory: 'E:\Git\nas2dfork\test\x86\Debug'
Command produced no output
[5/23/2019 4:58:56 PM Warning] No test is available in E:\Git\nas2dfork\proj\vs2019\Win32\Debug\NAS2D_d.lib E:\Git\nas2dfork\test\x86\Debug\test.exe. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
[5/23/2019 4:58:56 PM Informational] ========== Discover test finished: 0 found (0:00:00.9179184) ==========
[5/23/2019 4:59:09 PM Informational] ------ Run test started ------
[5/23/2019 4:59:10 PM Warning] Could not locate debug symbols for 'E:\Git\nas2dfork\test\x86\Debug\test.exe'. To make use of '--list_content' discovery, ensure that debug symbols are available or make use of '<ForceListContent>' via a .runsettings file.
[5/23/2019 4:59:10 PM Informational] Test Adapter for Google Test: Test execution starting...
[5/23/2019 4:59:10 PM Error] ERROR: Could not list test cases for executable 'E:\Git\nas2dfork\test\x86\Debug\test.exe': process execution failed with exit code -1073741515
Command executed: 'E:\Git\nas2dfork\test\x86\Debug\test.exe --gtest_list_tests', working directory: 'E:\Git\nas2dfork\test\x86\Debug'
Command produced no output
[5/23/2019 4:59:10 PM Informational] Found 0 tests in executable E:\Git\nas2dfork\test\x86\Debug\test.exe
[5/23/2019 4:59:10 PM Informational] Google Test execution completed, overall duration: 00:00:00.1294859
[5/23/2019 4:59:10 PM Error] ERROR: 
================
The following errors and warnings have occured during test execution  (enable debug mode for more information):
ERROR: Could not list test cases for executable 'E:\Git\nas2dfork\test\x86\Debug\test.exe': process execution failed with exit code -1073741515
Command executed: 'E:\Git\nas2dfork\test\x86\Debug\test.exe --gtest_list_tests', working directory: 'E:\Git\nas2dfork\test\x86\Debug'
Command produced no output
[5/23/2019 4:59:10 PM Warning] No test is available in E:\Git\nas2dfork\proj\vs2019\Win32\Debug\NAS2D_d.lib E:\Git\nas2dfork\test\x86\Debug\test.exe. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
[5/23/2019 4:59:10 PM Informational] ========== Run test finished: 0 run (0:00:01.1851544) ==========
[5/23/2019 5:01:25 PM Informational] ------ Discover test started ------
[5/23/2019 5:01:26 PM Warning] Could not locate debug symbols for 'E:\Git\nas2dfork\test\x86\Release\test.exe'. To make use of '--list_content' discovery, ensure that debug symbols are available or make use of '<ForceListContent>' via a .runsettings file.
[5/23/2019 5:01:26 PM Informational] Test Adapter for Google Test: Test discovery starting...
[5/23/2019 5:01:26 PM Informational] Found 17 tests in executable E:\Git\nas2dfork\test\x86\Release\test.exe
[5/23/2019 5:01:26 PM Informational] Test discovery completed, overall duration: 00:00:00.1719052
[5/23/2019 5:01:26 PM Informational] ========== Discover test finished: 17 found (0:00:00.9617187) ==========
[5/23/2019 5:01:52 PM Informational] ------ Discover test started ------
[5/23/2019 5:01:53 PM Warning] Could not locate debug symbols for 'E:\Git\nas2dfork\test\x86\Debug\test.exe'. To make use of '--list_content' discovery, ensure that debug symbols are available or make use of '<ForceListContent>' via a .runsettings file.
[5/23/2019 5:01:53 PM Informational] Test Adapter for Google Test: Test discovery starting...
[5/23/2019 5:01:53 PM Error] ERROR: Could not list test cases for executable 'E:\Git\nas2dfork\test\x86\Debug\test.exe': process execution failed with exit code -1073741515
Command executed: 'E:\Git\nas2dfork\test\x86\Debug\test.exe --gtest_list_tests', working directory: 'E:\Git\nas2dfork\test\x86\Debug'
Command produced no output
[5/23/2019 5:01:53 PM Informational] Found 0 tests in executable E:\Git\nas2dfork\test\x86\Debug\test.exe
[5/23/2019 5:01:53 PM Informational] Test discovery completed, overall duration: 00:00:00.1073571
[5/23/2019 5:01:53 PM Error] ERROR: 
================
The following errors and warnings have occured during test discovery  (enable debug mode for more information):
ERROR: Could not list test cases for executable 'E:\Git\nas2dfork\test\x86\Debug\test.exe': process execution failed with exit code -1073741515
Command executed: 'E:\Git\nas2dfork\test\x86\Debug\test.exe --gtest_list_tests', working directory: 'E:\Git\nas2dfork\test\x86\Debug'
Command produced no output
[5/23/2019 5:01:53 PM Warning] No test is available in E:\Git\nas2dfork\proj\vs2019\Win32\Debug\NAS2D_d.lib E:\Git\nas2dfork\test\x86\Debug\test.exe. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
[5/23/2019 5:01:53 PM Informational] ========== Discover test finished: 0 found (0:00:00.8351038) ==========
[5/23/2019 5:02:00 PM Informational] ------ Discover test started ------
[5/23/2019 5:02:01 PM Warning] Could not locate debug symbols for 'E:\Git\nas2dfork\test\x86\Debug\test.exe'. To make use of '--list_content' discovery, ensure that debug symbols are available or make use of '<ForceListContent>' via a .runsettings file.
[5/23/2019 5:02:01 PM Informational] Test Adapter for Google Test: Test discovery starting...
[5/23/2019 5:02:01 PM Error] ERROR: Could not list test cases for executable 'E:\Git\nas2dfork\test\x86\Debug\test.exe': process execution failed with exit code -1073741515
Command executed: 'E:\Git\nas2dfork\test\x86\Debug\test.exe --gtest_list_tests', working directory: 'E:\Git\nas2dfork\test\x86\Debug'
Command produced no output
[5/23/2019 5:02:01 PM Informational] Found 0 tests in executable E:\Git\nas2dfork\test\x86\Debug\test.exe
[5/23/2019 5:02:01 PM Informational] Test discovery completed, overall duration: 00:00:00.1291379
[5/23/2019 5:02:01 PM Error] ERROR: 
================
The following errors and warnings have occured during test discovery  (enable debug mode for more information):
ERROR: Could not list test cases for executable 'E:\Git\nas2dfork\test\x86\Debug\test.exe': process execution failed with exit code -1073741515
Command executed: 'E:\Git\nas2dfork\test\x86\Debug\test.exe --gtest_list_tests', working directory: 'E:\Git\nas2dfork\test\x86\Debug'
Command produced no output
[5/23/2019 5:02:01 PM Warning] No test is available in E:\Git\nas2dfork\proj\vs2019\Win32\Debug\NAS2D_d.lib E:\Git\nas2dfork\test\x86\Debug\test.exe. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
[5/23/2019 5:02:01 PM Informational] ========== Discover test finished: 0 found (0:00:00.9874947) ==========
[5/23/2019 5:02:05 PM Informational] ------ Run test started ------
[5/23/2019 5:02:06 PM Warning] Could not locate debug symbols for 'E:\Git\nas2dfork\test\x86\Debug\test.exe'. To make use of '--list_content' discovery, ensure that debug symbols are available or make use of '<ForceListContent>' via a .runsettings file.
[5/23/2019 5:02:06 PM Informational] Test Adapter for Google Test: Test execution starting...
[5/23/2019 5:02:06 PM Error] ERROR: Could not list test cases for executable 'E:\Git\nas2dfork\test\x86\Debug\test.exe': process execution failed with exit code -1073741515
Command executed: 'E:\Git\nas2dfork\test\x86\Debug\test.exe --gtest_list_tests', working directory: 'E:\Git\nas2dfork\test\x86\Debug'
Command produced no output
[5/23/2019 5:02:06 PM Informational] Found 0 tests in executable E:\Git\nas2dfork\test\x86\Debug\test.exe
[5/23/2019 5:02:06 PM Informational] Google Test execution completed, overall duration: 00:00:00.1204878
[5/23/2019 5:02:06 PM Error] ERROR: 
================
The following errors and warnings have occured during test execution  (enable debug mode for more information):
ERROR: Could not list test cases for executable 'E:\Git\nas2dfork\test\x86\Debug\test.exe': process execution failed with exit code -1073741515
Command executed: 'E:\Git\nas2dfork\test\x86\Debug\test.exe --gtest_list_tests', working directory: 'E:\Git\nas2dfork\test\x86\Debug'
Command produced no output
[5/23/2019 5:02:06 PM Warning] No test is available in E:\Git\nas2dfork\proj\vs2019\Win32\Debug\NAS2D_d.lib E:\Git\nas2dfork\test\x86\Debug\test.exe. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
[5/23/2019 5:02:06 PM Informational] ========== Run test finished: 0 run (0:00:00.8908809) ==========

**Comment by DanRStevens** [comment]
Ahh, instead of modifying the additional library names, try updating additional library paths. That way things still work if the libraries are found in another place within the library search path. In particular, libraries can still be found if they are installed through another package manager, or some other means.

Though I do wonder if vcpkg is setup correctly for your local environment. Assuming you've installed Google Test with vcpkg, I assume it should set an additional library path for you automatically. That should be how AppVeyor manages to find those libraries, so I assume it should work for you. As it seems to not be working for you, I suspect something is wrong with the vcpkg integration when building the test project.

----

I'm not too sure what you're describing in the later part of your message. Though you said the main project doesn't build. If the main project doesn't build, there will be nothing for the test project to link with. Hence the test project also shouldn't build. The test project might compile, but it won't link unless the main static library project has built (compile + link).

If tests are not running properly, or there is test discovery failures, it could be because of an earlier problem while building the main project. I think that part needs to be addressed first.

I'm not too sure about the code analysis settings. These should be irrelevant for the test project. Perhaps make sure the test project is dependent on the appropriate configuration of the main project. Generally I would test against the release version of the static library. If the test project has both Debug and Release settings, they should depend on the corresponding configurations of the main project.

**Comment by DanRStevens** [comment]
I'm guess this was resolved a long time ago and should be closed now.

**Comment by cugone** [comment]
Sort of. If I try to build in Debug the gtest.lib isn't found. Release works just fine though.

**Comment by DanRStevens** [comment]
I just verified this in a branch, using AppVeyor. We do indeed have a project configuration problem for the `Debug` build.

**Comment by DanRStevens** [comment]
I found some relevant information here:
[vcpkg does not work for google test](https://stackoverflow.com/questions/41315739/vcpkg-does-not-work-for-google-test)

I wanted to avoid specifying a library path and name together though, since they are normally specified separately. I also had a hunch it was about the naming of the library, since debug libraries often have "d" appended to the name.

After some experimenting, I found it was indeed the library name. For the debug configurations, I renamed the additional libraries:
- `gtest.lib` => `gtestd.lib`
- `gtest_main.lib` => `gtest_maind.lib`
